[Merged by Bors] - feat(field_theory/finite): Chevalley–Warning

jcommelin opened the pull request:
---

jcommelin committed:
Start on Chevalley-Warning

jcommelin committed:
WIP

jcommelin committed:
WIP

jcommelin committed:
Fix two sorries

jcommelin committed:
Annoying stuff is left

jcommelin committed:
Finish proof of Chevalley-Warning

jcommelin committed:
Fix build, and clean up

jcommelin committed:
Little change

sgouezel commented on src/data/mv_polynomial.lean:
```diff
@@ -660,32 +672,59 @@ finset.sup_le $ assume n hn,
+--   { simp only [nat.nat_zero_eq_zero, zero_mul, pow_zero, total_degree_one] },
+--   { rw [pow_succ, total_degree_mul, ih, nat.succ_mul, add_comm] },
+-- end
+
```
Did you mean to keep these?

jcommelin replied:
Aaah, good point. I'll remove them.

jcommelin committed:
Remove commented code

jcommelin committed:
Undo renaming

jcommelin committed:
Oops, forgot one

jcommelin committed:
Reorganise files

jcommelin committed:
Fix build

bryangingechen commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+-/
+
+import data.equiv.algebra
+import algebra.char_p
```
Merging `field_theory.finite_card` with `field_theory.finite` breaks `data.zmod.quadratic_reciprocity` as in https://github.com/leanprover-community/mathlib/pull/819#issuecomment-472830821 since `algebra.char_p` imports `data.padics.padic_norm` which [has `attribute [class] nat.prime`](https://github.com/leanprover-community/mathlib/blob/81a31ca4a8c0287bf0b0ce40f1a0df16543b7abe/src/data/padics/padic_norm.lean#L48).

bryangingechen replied:
Oops, I commented on the wrong commit, but the above holds true at 4f7e38a too.

jcommelin replied:
This is now fixed.

rwbarton commented on src/algebra/big_operators.lean:
```diff
@@ -504,6 +504,21 @@ begin
+@[move_cast]
+lemma nat_cast_sum (f : α → ℕ) :
+  ((s.sum f : ℕ) : β) = s.sum (λ i, (f i : β)) :=
+(finset.sum_hom _).symm
```
These seem to be the same as `sum_nat_cast`, `prod_nat_cast` in the same file. Unless a different coercion is involved...?

jcommelin replied:
Fixed.

rwbarton commented on src/data/fintype.lean:
```diff
@@ -179,6 +179,20 @@ def of_subsingleton (a : α) [subsingleton α] : fintype α :=
+open_locale classical
+
+@[to_additive]
+lemma prod_extend_by_one [comm_monoid α] {ι} [fintype ι] (s : finset ι) (f : ι → α) :
```
I'm inclined to say that you should instead write
```suggestion
lemma prod_extend_by_one [comm_monoid α] {ι} [fintype ι] [decidable_eq ι] (s : finset ι) (f : ι → α) :
```
and remove `open_locale classical`. However I'm not sure that won't break your usage! @ChrisHughes24 any thoughts here?

jcommelin replied:
Done.

ChrisHughes24 replied:
I only just saw this. I'm inclined to say the same as Reid, otherwise there'll be some diamond, and the `rw` won't work.

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+begin
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
```
Cute proof, but you actually only use `hq` where the weaker inequality `1 <= fintype.card K` is needed. There `exact fintype.card_pos_iff.mpr ⟨0⟩` is enough. (Maybe worth adding a lemma `fintype.one_le_card_iff := fintype.card_pos_iff`.)

jcommelin replied:
Thanks, fixed.

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
+  cases is_cyclic.exists_generator (units K) with a ha,
```
You could write
```lean
let ⟨a, ha⟩ := is_cyclic.exists_generator (units K) in
calc ...
```
and avoid "fake tactic mode".

jcommelin replied:
Mwah... there is so much tactic mode in this proof that I don't care about a little bit of fake tactic mode as well. (I know that `calc` is a fake tactic. But all the `begin ... end` blocks aren't)
Anyway, I can fix it if you prefer.

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+        nat.cast_sub, nat.cast_one],
+      exact le_trans hq (nat.pred_le _), },
+    { have key := geom_sum_mul (a^i : K) (q-1),
+      have hai : (a^i : K) ≠ 0, { rw ← units.coe_pow, apply units.ne_zero },
```
ohai!

jcommelin replied:
hello! Thanks for reading this proof in such detail!

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+      have hai : (a^i : K) ≠ 0, { rw ← units.coe_pow, apply units.ne_zero },
+      rw [pow_card_sub_one_eq_one _ hai, sub_self] at key,
+      replace key := eq_zero_or_eq_zero_of_mul_eq_zero key,
+      rw classical.or_iff_not_imp_right at key, apply key, contrapose! H,
```
(GitHub isn't smart enough to know I want to replace two lines here)
```suggestion
      apply (eq_zero_or_eq_zero_of_mul_eq_zero key).resolve_right,
      contrapose! H,
```

jcommelin replied:
Done

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+    exact ⟨1⟩ },
+  by_cases hi : i = 0,
+  { rcases char_p.exists K with ⟨p, _char_p⟩, resetI,
+    rcases card K p with ⟨n, hp, hn⟩,
```
These two `rcases` are unnecessary.

jcommelin replied:
Fixed, thanks.

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+  rw [← sum_sdiff (subset_univ (finset.singleton (0:K))), sum_singleton,
+    zero_pow (nat.pos_of_ne_zero hi), add_zero],
+  have := sum_pow_units K i,
+  have not_dvd_i : ¬q - 1 ∣ i,
```
Proof can be `{ contrapose! h, exact nat.le_of_dvd (nat.pos_of_ne_zero hi) h },`.

jcommelin replied:
Done, thanks.

rwbarton commented on src/field_theory/finite/basic.lean:
```diff
@@ -0,0 +1,205 @@
+begin
+  have hq : 0 < q - 1,
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
```
This isn't needed after other suggested changes.

jcommelin replied:
Still a TODO

rwbarton commented on src/field_theory/finite/chevalley_warning.lean:
```diff
@@ -0,0 +1,176 @@
+    rw [sum_const, nat.smul_eq_mul, mul_comm, card_univ] at this,
+    rwa [finsupp.sum, show d.support = univ, from _],
+    rw eq_univ_iff_forall,
+    intro i, rw [finsupp.mem_support_iff, ne.def, ← nat.le_zero_iff],
```
```suggestion
    intro i, rw [finsupp.mem_support_iff, ← nat.pos_iff_ne_zero],
```

jcommelin replied:
Done thanks

rwbarton commented on src/field_theory/finite/chevalley_warning.lean:
```diff
@@ -0,0 +1,176 @@
+    rwa [finsupp.sum, show d.support = univ, from _],
+    rw eq_univ_iff_forall,
+    intro i, rw [finsupp.mem_support_iff, ne.def, ← nat.le_zero_iff],
+    push_neg, exact lt_of_lt_of_le hq (h _), },
```
```suggestion
    exact lt_of_lt_of_le hq (h _), },
```

jcommelin replied:
Fixed.

rwbarton commented on src/field_theory/finite/chevalley_warning.lean:
```diff
@@ -0,0 +1,176 @@
+    rw eq_univ_iff_forall,
+    intro i, rw [finsupp.mem_support_iff, ne.def, ← nat.le_zero_iff],
+    push_neg, exact lt_of_lt_of_le hq (h _), },
+  by_cases hd' : d.support = univ,
```
I think the proof gets too complicated starting here. As far as I can see, you only introduce `hd'` in order to rewrite a product over `d.support` to a product over `univ`. But, the factors of the product over `univ` not belonging to `d.support` would be `1` anyways. So you should be able to do the rewrite unconditionally, using something like your lemma `prod_extend_by_one`, and then drop the entire second half of the remaining proof.

jcommelin replied:
Still a TODO, but Github thinks it's "outdated" because the code moved.

jcommelin replied:
I've tried to refactor the proof. I think it did not get shorter, but it is conceptually better structured. I'll see if I can factor out useful sublemmas.

jcommelin replied:
This is now done

rwbarton commented on src/field_theory/finite/chevalley_warning.lean:
```diff
@@ -0,0 +1,176 @@
+  { rw [← card_units, fintype.card_pos_iff],
+    exact ⟨1⟩ },
+  rcases char_p.exists K with ⟨p, _char_p⟩, resetI,
+  rcases card K p with ⟨n, hp, hn⟩,
```
These aren't used.

jcommelin replied:
Fixed, thanks.

rwbarton commented on src/field_theory/finite/chevalley_warning.lean:
```diff
@@ -0,0 +1,176 @@
+    refine le_trans _ (finset.le_sup hd),
+    have : univ.sum (λ s:σ, q-1) ≤ univ.sum (λ s, d s) := sum_le_sum (λ s _, h s),
+    rw [sum_const, nat.smul_eq_mul, mul_comm, card_univ] at this,
+    rwa [finsupp.sum, show d.support = univ, from _],
```
Like below, after `rw finsupp.sum, convert this using 1`, you have another application of `sum_extend_by_zero` or something like it. So you can ignore the suggested changes below, and again you don't actually need `hq`.

jcommelin replied:
Still a TODO, but Github thinks it's "outdated" because the code moved.

jcommelin replied:
Done.

rwbarton commented on src/field_theory/finite/chevalley_warning.lean:
```diff
@@ -0,0 +1,176 @@
+      intros x hx, clear hx,
+      rw show F.eval x = finset.prod s (λ (i : ι), (1 - f i ^ (q - 1)).eval x),
+      { convert eval₂_prod _ _ _ _, exact is_semiring_hom.id },
+      split_ifs with hx hx,
```
You could make the rest of this proof cleaner by defining a lemma `a ^ (q - 1) = if a = 0 then 0 else 1`, then rewriting inside the product with `conv` (before doing `split_ifs`), then writing a lemma for multiplying a bunch of `if p i then 0 else 1`s together, if it doesn't already exist.

jcommelin replied:
Still a TODO, but Github thinks it's "outdated" because the code moved.

jcommelin replied:
Done

rwbarton reviewed (commented):
Oof, I underestimated how sticky this situation got. Sorry, I'm partly responsible for this mess. At this point, I think it is best to proceed in small steps. I suggest not doing any refactoring or variable renaming in this PR, and just adding the new content to a new file. I think the new content in `field_theory.finite.basic` is `sum_pow_units`, `cast_card_eq_zero`, `sum_pow_lt_card_sub_one`; probably you can verify this easily.

Since I'm already in the process of looking at this, I'll review it as though this has already happened.

jcommelin committed:
Merge branch 'master' into chev-warn

jcommelin committed:
Fix build

jcommelin committed:
WIP

jcommelin committed:
Fix build

jcommelin committed:
Implement some suggestions by Reid

jcommelin committed:
Remove field_theory/finite/

jcommelin committed:
Use fact

jcommelin committed:
Refactoring a bunch of proofs, and linting

jcommelin committed:
Minor cleanups

jcommelin committed:
Refactor

jcommelin committed:
Prove a version of Chev-Warn for 1 polynomial

jcommelin committed:
Linting

jcommelin committed:
Linting

jcommelin committed:
More refactoring

jcommelin committed:
Move things to proper place

ChrisHughes24 commented on src/group_theory/order_of_element.lean:
```diff
@@ -368,6 +372,30 @@ calc (univ.filter (λ a : α, a ^ n = 1)).card ≤ (gpowers (g ^ (fintype.card 
+
+variables [group α]
+
+lemma is_cyclic.gpowers_eq_univ (ha : ∀ x : α, x ∈ gpowers a) :
```
This isn't really about cyclic groups. `gpowers a` could be replaced with any set and the lemma still holds. Same with `is_cyclic.gpowers_to_finset_eq_univ`

jcommelin replied:
Good point :smile:

jcommelin committed:
Remove some lemmas that are already there in greater generality

jcommelin committed:
fix build

jcommelin committed:
Minor tweak

jcommelin committed:
feat(*): small additions that prepare for Chevalley-Warning

jcommelin committed:
Merge branch 'chev-warn-pre1' into chev-warn

jcommelin committed:
refactor proof

jcommelin committed:
More refactoring

jcommelin committed:
Small change

jcommelin committed:
More refactoring

jcommelin committed:
don't use fintype.prod/sum

jcommelin committed:
Fix notation

jcommelin committed:
Merge branch 'master' into chev-warn

jcommelin committed:
Merge branch 'master' into chev-warn

jcommelin committed:
refactor(field_theory): merge finite_card into finite

Now that field_theory.finite_card no longer depends on padic_norm
we can merge it into field_theory.finite, where it belongs.

This PR also adds some preparations for the Chevalley–Warning theorem.

jcommelin committed:
Update src/field_theory/finite.lean

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

jcommelin committed:
Apply suggestions from code review

Co-authored-by: sgouezel <sebastien.gouezel@univ-rennes1.fr>

jcommelin committed:
Refactor proof

jcommelin committed:
Add docstring

jcommelin committed:
Merge branch 'master' into finite-fields

jcommelin committed:
Use the fact that calc now refines

jcommelin committed:
Merge branch 'master' into finite-fields

jcommelin committed:
WIP

jcommelin committed:
chore(field_theory/finite): meaningful variable names

jcommelin committed:
refactor(field_theory): move finite_card.lean into finite.lean

jcommelin committed:
Ooops, forgot to actually delete the file

jcommelin committed:
Merge branch 'finite-card' into finite-fields

jcommelin committed:
Merge branch 'master' into finite-fields

jcommelin committed:
Merge branch 'master' into finite-fields

jcommelin committed:
Linting

jcommelin committed:
Slight golfing

jcommelin committed:
Merge branch 'finite-fields' into chev-warn

jcommelin committed:
Merge branch 'master' into chev-warn

jcommelin committed:
Fix build

jcommelin committed:
Undo open nat

jcommelin committed:
Minimise imports

jcommelin committed:
Move code to correct place

jcommelin committed:
Restructure proof

bryangingechen reviewed (commented):
This looks great! A few questions: 
- I could be wrong, but I think these theorems are more likely to be applied to polynomials than to finite fields. Would it make sense for these theorems to be under the `mv_polynomial` namespace?
- It seems to me like `f`, `d` and `s` (in the various statements here) can be inferred from later arguments. Should these be made implicit? (Of course if the answer to the question above is yes, then `f : mv_polynomial σ K` should remain explicit.)

jcommelin committed:
Fix implicit/explicit variables and namespaces

bryangingechen commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+variables {K : Type*} {σ : Type*} [fintype K] [field K] [fintype σ]
+local notation `q` := fintype.card K
+
+lemma mv_polynomial.exists_degree_lt_card_sub_one (f : mv_polynomial σ K)
```
Why not just wrap the whole file in `namespace mv_polynomial`? (Any particular reason why the two main `theorem`s should not be namespaced?)

jcommelin replied:
They felt too specific. Also, the family version doesn't take an `mv_polynomial` as argument, so you can't use dot notation. But I don't have a strong opinion on this.

bryangingechen replied:
Ah, OK. I was originally thinking that more specific statements should go in more specific namespaces, but on reflection, the more specific a statement is, the less likely there will be name conflicts, so leaving it in the root namespace may just lead to less overall verbosity.

bryangingechen commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -121,9 +120,9 @@ in finitely many variables (`X s`, `s : σ`) over a finite field of characterist
 Then the number of common solutions of the `f i` is divisible by `p`. -/
 theorem char_dvd_card_solutions_family (p : ℕ) [char_p K p]
-  {ι : Type*} (s : finset ι) (f : ι → (mv_polynomial σ K))
+  {ι : Type*} {s : finset ι} (f : ι → (mv_polynomial σ K))
```
```suggestion
  {ι : Type*} {s : finset ι} {f : ι → mv_polynomial σ K}
```
This is inferrable from `h`, right?

bryangingechen commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -182,15 +181,15 @@ Assume that the total degree of `f` is less than the cardinality of `σ`.
 theorem char_dvd_card_solutions (p : ℕ) [char_p K p]
-  (f : mv_polynomial σ K) (h : f.total_degree < fintype.card σ) :
-  (p:ℕ) ∣ fintype.card {x : σ → K // f.eval x = 0} :=
+  (f : mv_polynomial σ K) {h : f.total_degree < fintype.card σ} :
```
```suggestion
  {f : mv_polynomial σ K} (h : f.total_degree < fintype.card σ) :
```
Should it be this way instead?

bryangingechen commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -182,15 +181,15 @@ Assume that the total degree of `f` is less than the cardinality of `σ`.
-  { simpa only [fintype.univ_punit, sum_singleton] using h, }
+  have : ∑ i : unit, (F i).total_degree < fintype.card σ,
+  { simpa only [fintype.univ_punit, sum_singleton] using h, },
+  have key := char_dvd_card_solutions_family p F this,
```
```suggestion
  have key := char_dvd_card_solutions_family p this,
```
(If you take the suggestion above.)

jcommelin committed:
Apply suggestions from code review

Co-authored-by: Bryan Gin-ge Chen <bryangingechen@gmail.com>

rwbarton commented on src/data/equiv/basic.lean:
```diff
@@ -838,6 +838,14 @@ def subtype_pi_equiv_pi {α : Sort u} {β : α → Sort v} {p : Πa, β a → Pr
+instance subtype.not_not_unique {α : Type*} (p : α → Prop) [decidable_pred p] [unique {x // p x}] :
+  unique {x // ¬ ¬ p x} :=
+(equiv.subtype_congr_right $ λ a, not_not).unique_of_equiv ‹_›
+
```
I don't care for these instances, particularly the second one:
* They're in a module unrelated to anything that appears in their types, where there's no reason to look for them, and which one might not have imported.
* Especially the second has a form that isn't likely to be guessed. But even with the first, it's not hard to imagine some other unrelated file having an instance of `unique {x // a = x}`.

Using these instances creates an implicit long-range dependency. You know that on line 86 of such-and-such a file you need these instances, but that information can't be recovered from the code except by looking at the instance search trace.

Better would be if these were `def`s somehow, but I think even better than that is to replace these by the `let e : K ≃ {x // x ∘ coe = x₀}` that you build in the course of the other proof. That seems also potentially useful in other situations (multilinear functions)?

jcommelin replied:
I think that I'm actually using them in the `prod_unique` a bit further down. But I agree with the general sentiment of your comment, and I'll see if I can refactor.

jcommelin replied:
Fixed.

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+  calc (q - 1) * fintype.card σ = ∑ s:σ, (q - 1) :
+                                  by rw [sum_const, nat.smul_eq_mul, mul_comm, card_univ]
+                            ... ≤ ∑ s, d s : sum_le_sum (λ s _, h s)
+                            ... ≤ d.sum (λ (n : σ) (e : ℕ), e) : by rwa [finsupp.sum, this]
```
I think you're working around a missing lemma for `finsupp` (at least, I did not find it), that says that when the domain of `d` is already a `fintype`, then `d.sum = \sum s, d s`. Then you could eliminate lines 52-54 and 56-59 of this proof I think.

jcommelin replied:
Thanks, done.

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+  calc (∑ x, f.eval x) = ∑ x : σ → K, ∑ d in f.support, f.coeff d * ∏ i, x i ^ d i : _
+                   ... = ∑ d in f.support, ∑ x : σ → K, f.coeff d * ∏ i, x i ^ d i : sum_comm
+                   ... = 0 : sum_eq_zero _,
+  { simp only [eval, eval₂, finsupp.sum, id.def, finsupp.prod_pow], refl, },
```
This should just be a lemma I think, for `f.eval` and for `f.eval₂`.

jcommelin replied:
Thanks, done.

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+            ∏ j, (x : σ → K) j ^ d j     : (fintype.sum_fiberwise _ _).symm
+                                 ... = 0 : fintype.sum_eq_zero _ _,
+  intros x₀,
+  let e : K ≃ {x // x ∘ coe = x₀} :=  (equiv.fun_unique _ _).symm.trans (equiv.subtype_preimage _ x₀).symm,
```
As mentioned above, this `e` should be a definition somewhere in `data.equiv.basic`.

jcommelin replied:
Do you have a suggestion for a name? Something like `equiv_fiber`?

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+      @fintype.prod_unique {j // j = i} _ _ _ (λ j, (e a : σ → K) j ^ d j),
+    rw ← this,
+    congr,
+  end
```
Maybe it would be clearer to move most or all of this proof to after the `calc`--it's not visually obvious whether line 103 resumes line 95 or line 91.

jcommelin replied:
Done

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+  let cd : K := f.coeff d,
+  calc (∑ x : σ → K, cd * ∏ i, x i ^ d i) = cd * (∑ x : σ → K, ∏ i, x i ^ d i) : mul_sum.symm
+                                      ... = 0 : (mul_eq_zero.mpr ∘ or.inr) _,
+  calc (∑ x : σ → K, ∏ i, x i ^ d i) =
```
I think it's easier to read when `=` starts the next line. Then the `=` on line 84 can match its alignment. Same on line 87 and line 94.

jcommelin replied:
Done

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+  have hF : ∀ x, F.eval x = if x ∈ S then 1 else 0,
+  { intro x,
+    calc F.eval x = ∏ i in s, (1 - f i ^ (q - 1)).eval x :
+      by { convert eval₂_prod id _ _ _, exact is_semiring_hom.id }
```
This should be (the missing) lemma `eval_prod`.

jcommelin replied:
Thanks, done

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+              ... = if x ∈ S then 1 else 0 : _,
+    simp only [eval_sub, eval_pow, eval_one],
+    split_ifs with hx hx,
+    { rw finset.prod_eq_one,
```
```suggestion
    { apply finset.prod_eq_one,
```

jcommelin replied:
Done

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+      by { convert eval₂_prod id _ _ _, exact is_semiring_hom.id }
+              ... = if x ∈ S then 1 else 0 : _,
+    simp only [eval_sub, eval_pow, eval_one],
+    split_ifs with hx hx,
```
Did anything ever come of the discussion about having a name for `if p then 1 else 0`?

jcommelin replied:
Not as far as I know...

robertylewis replied:
We use the name `boole` to refer to it in identifiers, but I don't think there's a definition by this name.

rwbarton commented on src/field_theory/chevalley_warning.lean:
```diff
@@ -0,0 +1,195 @@
+  show (1 - f i ^ (q - 1)).total_degree ≤ (q - 1) * (f i).total_degree,
+  calc (1 - f i ^ (q - 1)).total_degree
+          ≤ max (1 : mv_polynomial σ K).total_degree (f i ^ (q - 1)).total_degree : total_degree_sub _ _
+      ... ≤ (f i ^ (fintype.card K - 1)).total_degree :
```
```suggestion
      ... ≤ (f i ^ (q - 1)).total_degree :
```

rwbarton reviewed (commented):
Mostly I suggest factoring out more things into lemmas. At a slightly higher level, I also wonder whether it makes sense to turn this step of taking a sum of products, where only one factor is changing, and factoring out the changing part into a lemma somehow.

jcommelin committed:
Update src/field_theory/chevalley_warning.lean

Co-authored-by: Reid Barton <rwbarton@gmail.com>

jcommelin committed:
Process the easy comments

jcommelin committed:
Little fixes

jcommelin committed:
Linting

jcommelin committed:
Refactor an equiv

jcommelin committed:
Merge branch 'master' into chev-warn

jcommelin committed:
Linting

jcommelin committed:
Remove deps on awkard instances

jcommelin committed:
Move files to correct place in mathlib

bryangingechen committed:
Merge branch 'community-master' into chev-warn

bryangingechen committed:
Merge remote-tracking branch 'community/master' into chev-warn

bryangingechen committed:
fix build

bryangingechen committed:
Merge branch 'community-master' into chev-warn

bryangingechen committed:
fix build
